tecken/settings.py

```diff
@@ -573,42 +566,95 @@ def filter(self, record):
+).upper()
 
-# If UPLOAD_DEFAULT_URL is not in SYMBOL_URLS it's just too weird. It means we'd upload
-# to a S3 bucket we'd never read from and thus it'd be impossible to know the upload
-# worked.
-if UPLOAD_DEFAULT_URL not in SYMBOL_URLS:
-    raise ValueError(
-        f"The UPLOAD_DEFAULT_URL ({UPLOAD_DEFAULT_URL!r}) has to be one of the URLs "
-        f"in SYMBOL_URLS ({SYMBOL_URLS!r})"
+if CLOUD_SERVICE_PROVIDER == "GCS":
+    UPLOAD_GCS_BUCKET = _config(
+        "UPLOAD_GCS_BUCKET",
+        doc="The GCS bucket name for uploads and downloads.",
+    )
+    DOWNLOAD_S3_BUCKET = _config(
+        "DOWNLOAD_S3_BUCKET",
+        raise_error=False,
+        doc="The S3 bucket name for downloads.",
+    )
+    DOWNLOAD_S3_REGION = (
+        _config(
+            "DOWNLOAD_S3_REGION",
+            raise_error=False,
+            doc="The region of the S3 download bucket.",
+        )
+        or None
+    )
+    UPLOAD_BACKEND = {
+        "class": "tecken.ext.gcs.storage.GCSStorage",
+        "options": {
+            "bucket": UPLOAD_GCS_BUCKET,
+            "prefix": "v1",
+            "try_symbols": False,
+        },
+    }
+    TRY_UPLOAD_BACKEND = {
+        "class": "tecken.ext.gcs.storage.GCSStorage",
+        "options": {
+            "bucket": UPLOAD_GCS_BUCKET,
+            "prefix": "try/v1",
+            "try_symbols": True,
+        },
+    }
+    DOWNLOAD_BACKENDS = []
+    if DOWNLOAD_S3_BUCKET:
+        DOWNLOAD_BACKENDS = [
+            {
+                "class": "tecken.ext.s3.storage.S3Storage",
+                "options": {
+                    "bucket": DOWNLOAD_S3_BUCKET,
+                    "prefix": "v1",
+                    "try_symbols": False,
+                    "anonymous": True,
+                    "region": DOWNLOAD_S3_REGION,
+                },
+            },
+            {
+                "class": "tecken.ext.s3.storage.S3Storage",
+                "options": {
+                    "bucket": DOWNLOAD_S3_BUCKET,
+                    "prefix": "try/v1",
+                    "try_symbols": True,
+                    "anonymous": True,
+                    "region": DOWNLOAD_S3_REGION,
+                },
+            },
+        ]
+elif CLOUD_SERVICE_PROVIDER == "AWS":
+    UPLOAD_S3_BUCKET = _config(
+        "UPLOAD_S3_BUCKET",
+        doc="The S3 bucket name for uploads and downloads.",
+    )
+    UPLOAD_BACKEND = {
+        "class": "tecken.ext.s3.storage.S3Storage",
+        "options": {
+            "bucket": UPLOAD_S3_BUCKET,
+            "prefix": "v1",
+            "try_symbols": False,
+        },
+    }
+    TRY_UPLOAD_BACKEND = {
+        "class": "tecken.ext.s3.storage.S3Storage",
+        "options": {
+            "bucket": UPLOAD_S3_BUCKET,
+            "prefix": "try/v1",
+            "try_symbols": True,
+        },
+    }
+    DOWNLOAD_BACKENDS = []
+else:
+    raise ImproperlyConfigured(
+        f"invalid value for CLOUD_SERVICE_PROVIDER: {CLOUD_SERVICE_PROVIDER}"
```

Comment:
I wrote up https://github.com/willkg/everett/issues/253 to add a choices parser that would raise a configuration issue at the point of configuration when the value isn't one of an allowed list of values. We had a similar thing in Socorro and Antenna.

For now, we can leave it like this since it's temporary and we'll remove this once the GCP migration is over.

Reply:
That parser would be much nicer indeed.